### PR TITLE
replace IConsole with two TextWriters in Configuration

### PR DESCRIPTION
**Pull request by adamsitnik**
The whole idea behind this PR is to move `IConsole` to Rendering and extend the Config with two mutable `TextWriter` properties (`Out` and `Error`). Both properties are lazy initialized (perf), with default values of `Console.Out` and `Console.Error`. To test the produced output, they can be preconfigured with `StringWriter` instances. To disable the ouput, they can be set to `TextWriter.Null`.

```cs
/// <summary>
/// The standard output. Used by Help and other facilities that write non-error information.
/// By default it's set to <see cref="Console.Out"/>.
/// For testing purposes, it can be set to a new instance of <see cref="StringWriter"/>.
/// If you want to disable the output, please set it to <see cref="TextWriter.Null"/>.
/// </summary>
public TextWriter Out
{ 
    get => _output ??= Console.Out;
    set => _output = value ?? throw new ArgumentNullException(nameof(value), "Use TextWriter.Null to disable the output");
}

/// <summary>
/// The standard error. Used for printing error information like parse errors.
/// By default it's set to <see cref="Console.Error"/>.
/// For testing purposes, it can be set to a new instance of <see cref="StringWriter"/>.
/// </summary>
public TextWriter Error
{
    get => _error ??= Console.Error;
    set => _error = value ?? throw new ArgumentNullException(nameof(value), "Use TextWriter.Null to disable the output");
}
```

contributes to #1905, it's a first step towards making configuration mutable

### CODE REVIEW

File: src/System.CommandLine.ApiCompatibility.Tests/ApiCompatibilityApprovalTests.System_CommandLine_api_is_not_changed.approved.txt [hunk 1]
```diff
@@ -85,11 +80,13 @@ System.CommandLine
     public System.Collections.Generic.IReadOnlyList<Directive> Directives { get; }
     public System.Boolean EnablePosixBundling { get; }
     public System.Boolean EnableTokenReplacement { get; }
+    public System.IO.TextWriter Error { get; set; }
+    public System.IO.TextWriter Out { get; set; }
```
Comment (jonsequitur) [inline src/System.CommandLine.ApiCompatibility.Tests/ApiCompatibilityApprovalTests.System_CommandLine_api_is_not_changed.approved.txt:84]: The names `Error` and `Out` might need a little more context unless we're also renaming `CommandLineConfiguration`, e.g. `ConsoleOut` and `ConsoleError`.

**Review by jonsequitur — APPROVED** [review]

**Comment by tmds** [comment]
> The whole idea behind this PR is to move IConsole to Rendering and extend the Config with two mutable TextWriter properties (Out and Error).

```
    public interface IConsole :
        IStandardOut,
        IStandardError,
        IStandardIn
```

@adamsitnik @jonsequitur how does standard input fit into this?

**Comment by jonsequitur** [comment]
We never arrived at a satisfactory design for these interfaces. We decided it's safest to leave it out of scope for now.

**Comment by tmds** [comment]
That makes sense. My main rule of API design is _when in doubt, leave it out_.

This seems somewhat awkward:

```
parseResult.Configuration.Output.WriteLine("Hello");
```

**Comment by jonsequitur** [comment]
It is a little awkward. We're working on how to better separate parser configuration from invocation-related concerns. Suggestions are welcome.

**Pull request merged**